accepted/0000-automatic-migration-salt-bundle.md

```diff
@@ -0,0 +1,62 @@
+- Feature Name: Automatic migration from Salt 3000 to Salt Bundle
+- Start Date: 2023-03-22
+
+# Summary
+[summary]: #summary
+
+Force an automatic migration for Salt clients from Salt 3000 to Salt Bundle.
+
+
+# Motivation
+[motivation]: #motivation
+
+
+Salt 3000 is EOL as of Aug 31, 2022. It is advisable for Salt clients using Salt 3000 to migrate to Salt Bundle. Right now this migration can be done by applying the already existent `util.mgr_switch_to_venv_minion` state from the CLI. The puprose of this RFC is to automatically perform this migration for Salt 3000 clients to Salt Bundle so they can keep receiving updates for the Salt package, including security updates.
+
+
+# Detailed design
+[design]: #detailed-design
+
+An state to perform the migration from Salt 3000 to Salt Bundle is added to the highstate. This state checks that the 'saltversion' grain in the targeted minion and that the minion is not already running Salt Bundle. In that case, the migration is performed in the following steps:
+
+- Install the `venv-salt-minion` package.
+- Copy the minion id file, minion configuration files, PKI files and static grains for the current `salt-minion` in `/etc/salt` to the `venv-salt-minion` corresponding directories in `/etc/venv-salt-minion`.
+- Enable the `venv-salt-minion` service.
+- Disable the `salt-minion` service.
+
+The whole process is transparent to the user.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The `salt-minion` package remains installed and the configuration files in the `/etc/salt` are not deleted.
+
+- After migration this check is done in every highstate.
+
+- If a client was bootstrapped avoiding the Salt Bundle following the steps in the documentation, the migration would force this client to use Salt Bundle. Those steps are:
+
+  * Execute mgr-bootstrap with --no-bundle option.
+
+  * Set AVOID_VENV_SALT_MINION to 1 in the generated bootstrap script.
+
+  * For bootstrap state set the mgr_avoid_venv_salt_minion pillar to True.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Force the removal of the `salt-minion` package and related files. This could be done by means of a custom grain module that informs if the minion is running  `venv-salt-minion`. With that in place, the application of the highstate a second time could handle that second stage for purging.
+- The impact of not doing this is that Salt 3000 clients will stop receiving updates for the Salt package.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+- What is a better solution:
+  * Adding the sls to the highstate. The sls contains the logic for triggering the migration. 
```

Comment:
This would be the way that Salt is expected to be used. I know we don't really follow this approach, we just apply states and let them decide if there is something to do. The other option would be in line with our historic use of Salt states, even if it's not the idiomatic way.

Reply:
Am I missing something or was your comment actually targetting the other solution?

I'm don't a really have strong opinion here bu I would probably prefer to keep same approach than we currently use in SUMA (the logic is inside the SLS itself and we always add it to the highstate via our custom `mgr_master_tops` module).

Reply:
I think the sls containing the logic is a cleaner solution, when reading the sls file you get the whole picture and I believe is what Alex means when he says that it's hos Salt is expected to be used.

Reply:
I meant that the idiomatic way would be to have a state that describes the end situation, and then have this state applied to those systems that are targeted in top.sls. 

I'm okay with having the logic in the state to stay in line with our usage of SLS files.